### PR TITLE
[STACK iview2dcm -> refactor] Make iView to DICOM app

### PR DESCRIPTION
**Pull request by SimonBiggs**
Alright, so I am certainly stepping beyond my expertise here. Anyway... I think this is ready for review.

### CODE REVIEW

File: lib/pymedphys/_dicom/constants/core.py [hunk 1]
```diff
@@ -37,12 +37,6 @@
 
 from pymedphys._imports import pydicom
 
-# Many thanks to the Medical Connections for offering free
```
Comment (sjswerdloff) [inline lib/pymedphys/_dicom/constants/core.py:40]: Why are credits being deleted?   Was it moved?

File: lib/pymedphys/_experimental/streamlit/apps/iview_to_dicom.py [hunk 2]
```diff
@@ -72,19 +144,60 @@ def _create_portal_image_dicom_dataset(
...
+            # * is the image an ORIGINAL Image; an image whose pixel
+            #   values are based on original or source data
+            # * is the image a SECONDARY Image; an image created after
+            #   the initial patient examination
+            "ImageType": ["ORIGINAL", "SECONDARY", "PORTAL"],
+            # Purposefully avoided all concepts of 'Study' if possible
+            # as this would require appropriately grouping the images
+            # from each field. Certainly doable, and arguably it should
+            # be done, but out of scope for the current task.
+            "PatientID": patient_id,
```
Comment (sjswerdloff) [inline lib/pymedphys/_experimental/streamlit/apps/iview_to_dicom.py:169]: OK, this addresses a comment I made elsewhere.
Very good, happy to see it

File: lib/pymedphys/_experimental/streamlit/apps/iview_to_dicom.py [hunk 3]
```diff
@@ -72,19 +144,60 @@ def _create_portal_image_dicom_dataset(
...
+
+    # Many of the below was input without a full understanding of the
+    # implications of the choices being made.
     ds = _pp_dcm_create.dicom_dataset_from_dict(
         {
-            "ImageType": ["ORIGINAL", "PRIMARY", "PORTAL"],
+            "Modality": "RTIMAGE",
+            "InstanceCreatorUID": PYMEDPHYS_ROOT_UID,
+            "SOPClassUID": "1.2.840.10008.5.1.4.1.1.481.1",
+            "SOPInstanceUID": pydicom.uid.generate_uid(prefix=f"{PYMEDPHYS_ROOT_UID}."),
```
Comment (sjswerdloff) [inline lib/pymedphys/_experimental/streamlit/apps/iview_to_dicom.py:157]: Study and Series uids also.

File: lib/pymedphys/_experimental/streamlit/apps/iview_to_dicom.py [hunk 4]
```diff
@@ -72,19 +144,60 @@ def _create_portal_image_dicom_dataset(
...
+            "RTImageLabel": "",
+            "RTImagePlane": "NORMAL",
+            "PixelRepresentation": 0,
+            "XRayImageReceptorAngle": 0.0,
+            "RTImagePosition": None,
+            "PatientSupportAngle": table_angle,
+            "PixelData": pixel_array.tobytes(),
             "RadiationMachineSAD": sad,
             "RTImageSID": sid,
+            "PrimaryDosimeterUnit": "",
```
Comment (sjswerdloff) [inline lib/pymedphys/_experimental/streamlit/apps/iview_to_dicom.py:197]: MU should be fine.
If they are KV images, I'll need to look it up, but having something might be better than empty, and MU covers the vast majority of situations and systems

**Review by sjswerdloff — CHANGES_REQUESTED** [review]
Are you decompressing the jpeg?  Compatibility will be much better with uncompressed data.

Requested change is just to use MU for primary dosimeter unit

File: lib/pymedphys/_experimental/streamlit/apps/iview_to_dicom.py (hunk 2 above)
Comment (SimonBiggs) [reply lib/pymedphys/_experimental/streamlit/apps/iview_to_dicom.py:169]: > OK, this addresses a comment I made elsewhere.

Do you mean this comment:

https://github.com/pymedphys/pymedphys/pull/1387#issuecomment-774913098

Does that mean the `base` branch can be approved and merged?

**Comment by SimonBiggs** [comment]
> Are you decompressing the jpeg? Compatibility will be much better with uncompressed data.

Yup, that's what `pymedphys._losslessjpeg` is doing.

File: lib/pymedphys/_dicom/constants/core.py (hunk 1 above)
Comment (SimonBiggs) [reply lib/pymedphys/_dicom/constants/core.py:40]: Credits aren't being deleted. This was actually duplicated within the `constants/uuid.py` file. So, this duplication was deleted and the other file was used instead.

File: lib/pymedphys/_experimental/streamlit/apps/iview_to_dicom.py (hunk 3 above)
Comment (SimonBiggs) [reply lib/pymedphys/_experimental/streamlit/apps/iview_to_dicom.py:157]: Does the following address your concerns:

https://github.com/pymedphys/pymedphys/blob/ba2d8dc483b88e1a13d59623847f4038a3455127/lib/pymedphys/_experimental/streamlit/apps/iview_to_dicom.py#L165-L168

The same logic applies for series.

For now, it appears these files do import into pips. I don't want to create study or series UIDs at the moment. In the future I imagine these sorts of things could be passed in as a parameter to a more general function, and then that more general function could be called by a "study" or "series" creator function.

But... that's not anywhere within my scope at the moment.

File: lib/pymedphys/_experimental/streamlit/apps/iview_to_dicom.py (hunk 4 above)
Comment (SimonBiggs) [reply lib/pymedphys/_experimental/streamlit/apps/iview_to_dicom.py:197]: For this one, I'd prefer to err on the side of agreeing with what is currently exported from the iView database by the official tool, which, in this case is `""`.

**Comment by SimonBiggs** [comment]
@sjswerdloff when I believe I have addressed all concerns raised within the stack, did you want me to press the "request re-review" button to flag it for your attention?

image

**Comment by sjswerdloff** [comment]
> @sjswerdloff when I believe I have addressed all concerns raised within the stack, did you want me to press the "request re-review" button to flag it for your attention?
> 
> image

Yes

File: lib/pymedphys/_experimental/streamlit/apps/iview_to_dicom.py (hunk 4 above)
Comment (sjswerdloff) [reply lib/pymedphys/_experimental/streamlit/apps/iview_to_dicom.py:197]: > For this one, I'd prefer to err on the side of agreeing with what is currently exported from the iView database by the official tool, which, in this case is `""`.

I will submit a defect to the vendor

File: lib/pymedphys/_experimental/streamlit/apps/iview_to_dicom.py (hunk 2 above)
Comment (sjswerdloff) [reply lib/pymedphys/_experimental/streamlit/apps/iview_to_dicom.py:169]: Yes.
I think not providing the UID hierarchy early on in the construction 
is a mistake but it is your mistake to make.

**Review by sjswerdloff — APPROVED** [review]

**Comment by SimonBiggs** [comment]
> Yes

kk 🙂, will do. Just did it over at:

https://github.com/pymedphys/pymedphys/pull/1412

File: lib/pymedphys/_experimental/streamlit/apps/iview_to_dicom.py (hunk 2 above)
Comment (SimonBiggs) [reply lib/pymedphys/_experimental/streamlit/apps/iview_to_dicom.py:169]: Kk... if you feel strongly about this, I can see I have can expand on it appropriately on Thursday?

File: lib/pymedphys/_experimental/streamlit/apps/iview_to_dicom.py (hunk 4 above)
Comment (SimonBiggs) [reply lib/pymedphys/_experimental/streamlit/apps/iview_to_dicom.py:197]: 🙂

**Pull request merged**